parley/src/layout/cursor.rs

```diff
@@ -183,4 +233,250 @@ impl CursorPath {
     pub fn cluster<'a, B: Brush>(&self, layout: &'a Layout<B>) -> Option<Cluster<'a, B>> {
         self.run(layout)?.get(self.cluster_index)
     }
+
+    pub fn visual_line<'a, B: Brush>(&self, layout: &'a Layout<B>) -> Option<Line<'a, B>> {
+        layout.get(self.visual_line_index)
+    }
+}
+
+/// Returns a point that is falls within the vertical bounds of the given line.
+fn line_y<B: Brush>(line: &Line<B>) -> f32 {
+    line.metrics().baseline - line.metrics().ascent * 0.5
+}
+
+#[derive(Copy, Clone, PartialEq, Default, Debug)]
+pub struct Selection {
+    anchor: Cursor,
+    focus: Cursor,
+    h_pos: Option<f32>,
+}
```

Comment:
Should this be handling affinity?

Reply:
Affinity is a property of the cursor but in this code it’s not exposed directly. The two cases where this applies (mixed direction boundaries and end of line) are handled by generating an appropriate visual location for the cursor. All actual buffer manipulation is affinity agnostic.

Reply:
Oh apologies - I misread this

I had autocorrected the `Cursor` value as `usize`, because that's what it's always been before...